### PR TITLE
Implement a controller for ocs to ocs provider

### PR DESCRIPTION
**Pull request by iamniting**
create required resources to enable the remote connection on the provider.

Signed-off-by: Nitin Goyal <nigoyal@redhat.com>

### CODE REVIEW

**Comment by iamniting** [comment]
cc @nb-ohad @kesavanvt @sp98 

@jarrpa pls take a look

**Comment by iamniting** [comment]
Will remove `nigoyal` from logs once we are good with the approach as this helps me to test my PR easily.

File: controllers/storagecluster/generate.go
Comment (nb-ohad) [inline controllers/storagecluster/generate.go:84]: I think a better name for the deployment would be 'external-endpointorconsumer-endpoint`

Comment (iamniting) [reply controllers/storagecluster/generate.go:84]: Let's have more opinions on this. I am listing all 3 names here.

`ocs-storagecluster-external-endpoint`
`ocs-storagecluster-consumer-endpoint`
 `ocs-storagecluster-provider-server`

Comment (nb-ohad) [reply controllers/storagecluster/generate.go:84]: But why d we need `ocs-storagecluster` as a prefix? Even if we have multiple storage clusters we need a single server. So the options should be:
`ocs-external-endpoint`
`ocs-consumer-endpoint`
`ocs-provider-server`

Comment (iamniting) [reply controllers/storagecluster/generate.go:84]: correct I missed this.

File: controllers/storagecluster/server.go
Comment (iamniting) [inline controllers/storagecluster/server.go:21]: Question to the maintainers: How should I feed this image name in the ocs-operator?

File: controllers/storagecluster/generate.go
Comment (nb-ohad) [reply controllers/storagecluster/generate.go:84]: We might want to add `ocs-consumer-server` to the list

File: controllers/storagecluster/server.go
Comment (nb-ohad) [reply controllers/storagecluster/server.go:21]: I was under the impressions that we have decided to not have a separate image and that the image will be the ocs operator image (running from a different entry point)

Comment (iamniting) [reply controllers/storagecluster/server.go:21]: That is correct, But we still need to either feed the image or fetch it on runtime from ocs-operator deployment.

File: controllers/storagecluster/server.go
Comment (iamniting) [inline controllers/storagecluster/server.go:129]: Do we want to add owner references and monitor the resources for any changes? same goes for the service also.

File: controllers/storagecluster/server.go
Comment (iamniting) [inline controllers/storagecluster/server.go:181]: We can use node ports in range b/w `30000-32767`, Do we need any kind of approval before using any of the port?

File: controllers/storagecluster/server.go
Comment (nb-ohad) [reply controllers/storagecluster/server.go:21]: Why cant we just read the image name from the ocs-operator CSV?

Comment (nb-ohad) [reply controllers/storagecluster/server.go:21]: > We can do that but for that we either need a CSV name or we need to filter it out with ocs-operator name where ocs-operator deployment name will be always the same.

We already do exact that in the deployer, you can take a peak here:
https://github.com/red-hat-storage/ocs-osd-deployer/blob/fbc8169d7b7a81147309788813b07c2a88031aa1/controllers/managedocs_controller.go#L1389-L1399

> As far as I know, metrics exporter deployment is a part of the CSV which gets installed on the operator install not on storageCluster installation. So that's a different usecase.

Ok

Comment (iamniting) [reply controllers/storagecluster/server.go:21]: So you would like to filter with the `ocs-operator` prefix. ok I can do that

File: controllers/storagecluster/server.go
Comment (iamniting) [reply controllers/storagecluster/server.go:184]: I am not sure about that I can give it a try. @sp98 any idea?

File: controllers/storagecluster/server.go
Comment (nb-ohad) [reply controllers/storagecluster/server.go:21]: This is just one way to do it. Another option is to specify it in an env var as part of the CSV. That way it will be similar to all the other stuff that is already mentioned in the CSV.

File: controllers/storagecluster/server.go
Comment (jarrpa) [inline controllers/storagecluster/server.go:29]: I imagine all these log line will be removed before long. :wink:

File: controllers/storagecluster/server.go
Comment (jarrpa) [inline controllers/storagecluster/server.go:40]: ...don't use `obj` here, this is extremely confusing when the *struct you are writing a method for* is also `obj`. :stuck_out_tongue_closed_eyes: Honestly, though, you should change all the instances of `(obj *ocsProviderServer)` to something like `(srv *ocsProviderServer)`.

File: controllers/storagecluster/server.go
Comment (jarrpa) [reply controllers/storagecluster/server.go:132]: I'd say yes.

File: controllers/storagecluster/server.go
Comment (jarrpa) [reply controllers/storagecluster/server.go:184]: This is not required: https://v1-19.docs.kubernetes.io/docs/concepts/services-networking/service/#nodeport
```
      # Optional field
      # By default and for convenience, the Kubernetes control plane will allocate a port from a range (default: 30000-32767)
      nodePort: 30007
```

File: controllers/storagecluster/server.go
Comment (jarrpa) [reply controllers/storagecluster/server.go:21]: Could you just use downward API field references? https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/#the-downward-api e.g.:

```
      env:
        - name: IMAGE
          valueFrom:
            fieldRef:
              fieldPath: spec.containers[0].image
```

File: controllers/storagecluster/server.go
Comment (jarrpa) [inline controllers/storagecluster/server.go:46]: I might be forgetting, but I believe the `obj` being passed here would be overridden by `CreateOrUpdate()` with whatever it finds (or doesn't find). See the example here, where the desired PodTemplateSpec is defined separately and outside the `deploy` object: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate Thus, you must use this in-line function to enforce the state you want.

File: controllers/storagecluster/server.go
Comment (jarrpa) [inline controllers/storagecluster/server.go:62]: Why do you need to pass `r` and `instance` into this? I would imagine you should be able to just pass the Deployment you already have and do the calculation.

**Review by jarrpa — CHANGES_REQUESTED** [review]
All the use of "server" as a generic term for a very specific thing is confusing and technically misleading. Please rename `server.go` to something like `provider_server.go`. Similarly the instances of `apiServer` should probably be `ocsConsumerApiServer` or something similarly descriptive.

File: controllers/storagecluster/server.go
Comment (nb-ohad) [reply controllers/storagecluster/server.go:21]: @jarrpa @ issues here:
1. Referencing any container by an index is pruned to future issues due to container list location changes in future updates. 
2.  I don't think that the downward API support arbitrary spec fields, if I remember correctly there is a list of what can and can't be refernced to. But I might be wrong here.

File: controllers/storagecluster/server.go
Comment (nb-ohad) [reply controllers/storagecluster/server.go:46]: We use this pattern a lot in the deployer. 
The behaviour is a little bit different"
The object is taken then the result of the get is MERGED into it then the callback is run against the merged object to allow programmatic changes to the content before the create/update call is issued on the result.

The key point is that the merge operation will only affect fields that are defined in the k8s resource. This means that if the object presented to the function contains values that are undefined in the resource the outcome will be a struct that contains a merge of both.

File: controllers/storagecluster/server.go
Comment (iamniting) [reply controllers/storagecluster/server.go]: yes

File: controllers/storagecluster/server.go
Comment (iamniting) [reply controllers/storagecluster/server.go]: changed for loop obj to `resource` and method `obj` to `o`

File: controllers/storagecluster/server.go
Comment (iamniting) [reply controllers/storagecluster/server.go]: done

File: controllers/storagecluster/server.go
Comment (iamniting) [reply controllers/storagecluster/server.go]: removed

File: controllers/storagecluster/server.go
Comment (iamniting) [reply controllers/storagecluster/server.go]: I don't have the actual copy of the deployment, I have only the desired deployment. r is having a client which will help fetch the actual deployment with the help of an instance and verify its status.

File: controllers/storagecluster/generate.go
Comment (iamniting) [reply controllers/storagecluster/generate.go]: going with `ocs-provider-server`

**Comment by iamniting** [comment]
> All the use of "server" as a generic term for a very specific thing is confusing and technically misleading. Please rename `server.go` to something like `provider_server.go`. Similarly the instances of `apiServer` should probably be `ocsConsumerApiServer` or something similarly descriptive.

done

File: controllers/storagecluster/server.go
Comment (jarrpa) [reply controllers/storagecluster/server.go]: Looks like it. Okay, set the image as an annotation: https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#capabilities-of-the-downward-api

File: controllers/storagecluster/server.go
Comment (jarrpa) [reply controllers/storagecluster/server.go]: Sure. And that sounds like something we wouldn't want, right? If the Deployment we Get() has a different PodSpec, wouldn't that override the desired PodSpec you want to actually have?

File: controllers/storagecluster/server.go
Comment (jarrpa) [reply controllers/storagecluster/server.go]: I don't mean the desired Deployment, I mean the actual Deployment. That's what `resources` here gets set to, right? https://github.com/red-hat-storage/ocs-operator/pull/1421/files#diff-85e662b500fee0a341fbae01f02cd05d31663d6f6e355ab6536eb8b2ca0ea38fR53

File: controllers/storagecluster/server.go
Comment (nb-ohad) [reply controllers/storagecluster/server.go]: @jarrpa But why? what is the diff between setting it directly in the ENV var or setting an annotation and then referencing it in the env var? You still need to have it manually set up somewhere!

File: controllers/storagecluster/server.go
Comment (nb-ohad) [reply controllers/storagecluster/server.go]: Usually, you start with an empty obj! you should not think of object passed as the template
For the deployer we do the following:
1. create an "empty" obj
2. run `createOrUpdate`
3. have a callback that takes the content of the template and replace/merge it into the obj

An example: https://github.com/red-hat-storage/ocs-osd-deployer/blob/main/controllers/managedocs_controller.go#L599-L626

Comment (nb-ohad) [reply controllers/storagecluster/server.go]: @jarrpa we are in agreement here! the mutate function is the one to enforce desired state

Comment (iamniting) [reply controllers/storagecluster/server.go]: Yes we are, I had an discussion with Jose on the same

Comment (iamniting) [reply controllers/storagecluster/server.go]: done

File: controllers/storagecluster/server.go
Comment (iamniting) [reply controllers/storagecluster/server.go]: done

File: controllers/storagecluster/provider_server.go
Comment (sp98) [inline controllers/storagecluster/provider_server.go:27]: Where is `ensureCreated` being invoked?

**Review by sp98 — CHANGES_REQUESTED** [review]
Unit tests?

File: controllers/storagecluster/provider_server.go
Comment (iamniting) [reply controllers/storagecluster/provider_server.go:22]: yeah we are in discussion on this.

File: controllers/storagecluster/provider_server.go
Comment (iamniting) [reply controllers/storagecluster/provider_server.go:43]: I do want to reduce the reconcile calls wherever we can, thas why trying to create everything even any of them will fail. It should not affect us.

File: controllers/storagecluster/generate.go
Comment (iamniting) [reply controllers/storagecluster/generate.go:83]: yes will do that

File: controllers/storagecluster/provider_server.go
Comment (iamniting) [reply controllers/storagecluster/provider_server.go:209]: @jarrpa your opinion here?

File: controllers/storagecluster/generate.go
Comment (iamniting) [reply controllers/storagecluster/generate.go]: done

**Comment by iamniting** [comment]
> Unit tests?

Will create once we finalize the actual code. This will save efforts to change tests again and again.

File: controllers/storagecluster/provider_server.go
Comment (nb-ohad) [reply controllers/storagecluster/provider_server.go:172]: I think this is a good idea.

We have the same policy in the ocs-osd-deployer for most of the things that we reconcile. We do not use `embed` but `createOrUpdate` with a mutate function that starts with the template and applies dynamic changes on top of it. But the idea is almost identical.

File: controllers/storagecluster/provider_server.go
Comment (nb-ohad) [reply controllers/storagecluster/provider_server.go:213]: copying my reply from the last comment


> I think this is a good idea.

> We have the same policy in the ocs-osd-deployer for most of the things that we reconcile. We do not use embed but createOrUpdate with a mutate function that starts with the template and applies dynamic changes on top of it. But the idea is almost identical.

File: controllers/storagecluster/provider_server.go
Comment (jarrpa) [reply controllers/storagecluster/provider_server.go:172]: I'm not convinced that using `embed` would be an actual benefit to us... but I could be convinced otherwise.

Specifically, if we move all our static definitions into YAML files, we would then be in a place where we're parsing this text blob into a go struct. My main concern is how easy is it to debug errors? For example, what happens if we've changed a field or entire sub-struct in the go definition but the YAML manifest is not longer valid? Will that be any easier than continuing to hit nil-pointer errors because we keep using pointers to structs?

File: controllers/storagecluster/provider_server.go
Comment (jarrpa) [reply controllers/storagecluster/provider_server.go:47]: I think Nitin's approach is fine. As long as the underlying system won't be broken by it, I generally prefer to just blast out all Create/Update calls at once and let the system resolve itself over time. Here specifically, if two of these succeed but one doe snot due to something like a network hiccup, there's no sense in blocking on just the one failure.

File: controllers/storagecluster/provider_server.go
Comment (nb-ohad) [reply controllers/storagecluster/provider_server.go:172]: I would argue that YAML files are useless, the way we solved it in the deployer is by using const structs for the templates, one example can be found here: https://github.com/red-hat-storage/ocs-osd-deployer/blob/main/templates/alertmanagerconfig.go

This bring the benefits of static typing, and the ability to use code to generate some of the static parts

Comment (nb-ohad) [reply controllers/storagecluster/provider_server.go:172]: To complete the picture, the pattern we use for the reconcile is:
```
_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.alertmanagerConfig, func() error {
	if err := r.own(r.alertmanagerConfig); err != nil {
		return err
	}
	desired := templates.AlertmanagerConfigTemplate.DeepCopy()
	r.alertmanagerConfig.Spec = desired.Spec
	
	// Some code to apply the none static part of the reconcile
}

```

File: controllers/storagecluster/server.go
Comment (jarrpa) [reply controllers/storagecluster/server.go]: In general, hard-coding container images in code feels wrong. Unless there's a good reason to do it, the default should be to avoid it at all costs.

Here specifically, however, if we take the definition out of the go code and into the manifests that makes it easier for us to integrate it into our make/kustomize workflow so we don't have to worry about keeping track of it separately from what we're already doing to generate the bundle. It also gives us more flexibility as to where it can live, though off-hand if makes more sense (to me) to put the annotation on the Pod itself. Since the spec is generated by automation form the same value, I'm not to worried about writing it to multiple places.

File: controllers/storagecluster/provider_server.go
Comment (jarrpa) [reply controllers/storagecluster/provider_server.go:172]: Yeah, if we can let's do this. :point_up:

File: controllers/storagecluster/server.go
Comment (jarrpa) [reply controllers/storagecluster/server.go]: Ah-ha... just talked to Ohad and I realized I was thinking about the wrong thing. Still, literally because we have the whole kustomize automation process, we can (and probably should) just inject the container image name directly into the Pod's environment variable. Let's do that!

File: controllers/storagecluster/provider_server.go
Comment (iamniting) [reply controllers/storagecluster/provider_server.go:172]: We are doing something like this.

File: controllers/storagecluster/server.go
Comment (iamniting) [reply controllers/storagecluster/server.go]: done

File: controllers/storagecluster/provider_server.go
Comment (nb-ohad) [reply controllers/storagecluster/provider_server.go:173]: That is not the same:
1. The template is kept in memory but live for all reconciles not a single one
2. A template (as an env variable or a file) is more declarative by nature, as it is a definition a declaration of a definition,  while code  (as in a func) is less unless the entire func is a declaration of a variable that is immediately returned. But at that point, there is no reason they have the function in the first place.
3. If you have a function, you do not need a template as a function can take the actual object and mutate it, but then you lose the declarative nature of templates. 


To summarize, my opinion, on the matter:
- Copy from a template is a good solution
- Mutating using code is also a good solution (but less readable and prune to ordering errors) 
- Combining them is does NOT create something that is better then the original options

Comment (jarrpa) [reply controllers/storagecluster/provider_server.go:172]: I'm happy with this: https://github.com/red-hat-storage/ocs-operator/pull/1421/commits/bb44cbf3b72380f6aa1013d1c1d2834155e05d1e#diff-85e662b500fee0a341fbae01f02cd05d31663d6f6e355ab6536eb8b2ca0ea38fR91-R101

File: controllers/storagecluster/provider_server.go
Comment (jarrpa) [inline controllers/storagecluster/provider_server.go:101]: I'm happy with this! For me this resolves this discussion below: https://github.com/red-hat-storage/ocs-operator/pull/1421/commits/bb44cbf3b72380f6aa1013d1c1d2834155e05d1e#r765682700

Comment (nb-ohad) [reply controllers/storagecluster/provider_server.go:101]: Copying metadata is very problematic it removes additional labels and annotations that are set from the outside and this is something that we encounter again and again!

The metadata is both input and output we should never treat it as output-only or input only section

Comment (nb-ohad) [reply controllers/storagecluster/provider_server.go:101]: To clarify, I am referring to line 94

Comment (iamniting) [reply controllers/storagecluster/provider_server.go]: I agree, resolved the problem at all places.

**Comment by jarrpa** [comment]
Marking my general LGTM on this PR. It needs at least one more push to pass CI, but the fixes don't look like they'd change anything I'd disapprove. :wink:

File: controllers/storagecluster/provider_server.go
Comment (jarrpa) [inline controllers/storagecluster/provider_server.go:173]: I also like this! Look how simple it is. :grin: For a future PR, I would like to see the following:

* Rather than just returning `error`, I would suggest `bool, error`. Yes checking for `err != nil` achieves more or less the same thing, but for most any function starting with `is*` it makes semantic sense for it to be treated as or return a boolean. 
* In the error message, include what criteria weren't met for the Deployment to be considered "ready".

**Review by jarrpa — APPROVED** [review]

File: controllers/storagecluster/provider_server.go
Comment (nb-ohad) [reply controllers/storagecluster/provider_server.go:173]: Why `bool, error` for a sync function that works on memory only. I think `is` function should only return `bool` and the log or error should be created from the outside context. That way the util is more usable for other use cases where you don't actually need an error. For example, think about the `!is` use case

Comment (jarrpa) [reply controllers/storagecluster/provider_server.go:173]: > For example, think about the `!is` use case

Exactly this! Specifically, though, if the `is` function grows to more than just a single check (which it is now), we should understand why the function returned false when it does. And yes, we could effectively log the error in-place and return false, but then what does the human debugging / RCA experience look like? And error message gets propagated back up and could be exposed in the CR Status, but a log message is just a log. I think?

Comment (nb-ohad) [reply controllers/storagecluster/provider_server.go:173]: > if the is function grows to more than just a single check 
So it is not an `is` function anymore, if you insist you can also return a enum string for a reason. Again, no need for error

Comment (iamniting) [reply controllers/storagecluster/provider_server.go:173]: I changed the name of the func itself to the `ensureDeploymentReplica` than the func structure.

**Comment by iamniting** [comment]
@jarrpa Now CI is passing other than the `ocs-operator-ci`. It is working well on my machine. But somehow export is not working in the CI. Any idea what can be wrong here?

File: controllers/storagecluster/provider_server.go
Comment (sp98) [inline controllers/storagecluster/provider_server.go:22]:
```suggestion
	serverImage = GetEnvVarOrPanic("PROVIDER_API_SERVER_IMAGE")
```
Name is not used outside the package.

File: controllers/storagecluster/provider_server.go
Comment (sp98) [inline controllers/storagecluster/provider_server.go:26]:
```suggestion
	ocsProviderServerName = "ocs-provider-server"
```
Name is not used outside the package.

File: controllers/storagecluster/provider_server.go
Comment (sp98) [inline controllers/storagecluster/provider_server.go:29]: empty struct?

**Review by sp98 — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/provider_server.go
Comment (iamniting) [reply controllers/storagecluster/provider_server.go:22]: Not now, We may use them in the e2e tests which are not part of the same pkg.

File: controllers/storagecluster/provider_server.go
Comment (iamniting) [reply controllers/storagecluster/provider_server.go:26]: same as above.

File: controllers/storagecluster/provider_server.go
Comment (iamniting) [reply controllers/storagecluster/provider_server.go:29]: The struct is required to Implement the `resourceManager` Interface. https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/storagecluster/reconcile.go#L33

**Review by jarrpa — APPROVED** [review]
/lgtm

**Pull request merged**
